### PR TITLE
Implemented runtime serialization for monochrome images.

### PR DESCRIPTION
**Pull request by davidpiuva**
The ascii serialization is a lossy but compact way to store images in the source code.
It begins with the alphabet (a gradient from black to white) and continues with each line of pixels in the image. A pixel is represented by the character in the initial alphabet that is closest to the original luma value. Each line begins with `<` and ends with `>` so that text editors will not mess up the data that is stored using white space.

### CODE REVIEW

**Comment by marcusnaslund** [comment]
Hexadecimal stuff should probably reside in math.ooc or Numbers.ooc

**Comment by davidpiuva** [comment]
@fredrikbryntesson Peer review?

**Comment by fredrikbryntesson** [comment]
Kickin it to @marcusnaslund

File: source/draw/RasterMonochrome.ooc [hunk 1]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		output := 0.49f
+		for (rawValue in 0 .. 256) {
+			alphabetMap[rawValue] = alphabet[(output as Int) clamp(0, alphabet size - 1)]
+			output += scale
+		}
+		// Store alphabet
+		result append('<')
+		result append(alphabet)
+		result append('>')
+		result append('\n')
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: `result append('<') . append(alphabet) . append('>') . append('\n')`

File: source/draw/RasterMonochrome.ooc [hunk 2]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		String new(result)
+	}
+	// Parse an ascii image
+	fromAscii: static func (content: String) -> This {
+		// Measure dimensions and read the alphabet
+		alphabet: Char[128]
+		alphabetSize := 0
+		x := 0
+		y := -1
+		width := 0
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: I like `(alphabetSize, x, y, width) := (0, 0, -1, 0)` these days

File: source/draw/RasterMonochrome.ooc [hunk 3]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		inLine := false
+		for (i in 0 .. (content size)) {
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: This should work
```ooc
i := 0
while ((c := content[i]) != '\0') {
  i += 1
  ...
}
```
and would remove an indentation step

File: source/draw/RasterMonochrome.ooc [hunk 4]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+			raise(alphabetMap[code] > 0, "Character '" + alphabet[i] + "' (" + code toString() + ") is used more than once!")
+			value := ((i as Float) * (255.0f / ((alphabetSize - 1) as Float))) as Int clamp(0, 255)
+			alphabetMap[code] = value
+		}
+		// Allocate image
+		raise(width <= 0 || height <= 0, "An ascii image had zero dimensions!")
+		result := This new(IntVector2D new(width, height))
+		// Fill pixels
+		x = 0
+		y = -1
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: `(x, y) := (0, -1)`

File: source/draw/RasterMonochrome.ooc [hunk 5]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		inLine = false
+		for (i in 0 .. (content size)) {
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Same while-loop as above

File: source/draw/RasterMonochrome.ooc [hunk 6]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		result := CharBuffer new(((this size x + 3) * this size y) + 1)
+		// Generate mapping from luma to character
+		alphabetMap: Char[256]
+		scale := ((alphabet size - 1) as Float) / 255.0f
+		output := 0.49f
+		for (rawValue in 0 .. 256) {
+			alphabetMap[rawValue] = alphabet[(output as Int) clamp(0, alphabet size - 1)]
+			output += scale
+		}
+		// Store alphabet
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Not needed, but you can blank line here if you want.

File: source/draw/RasterMonochrome.ooc [hunk 7]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		result append('<')
+		result append(alphabet)
+		result append('>')
+		result append('\n')
+		// Serialize image
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Not needed

File: source/draw/RasterMonochrome.ooc [hunk 8]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
 		}
 		result
 	}
+	// Serialize to a lossy ascii image using an alphabet
+	// Precondition: alphabet may not have extended ascii, non printable, '\', '"', '>' or linebreak
+	// Example alphabet: " .,-_':;!+~=^?*abcdefghijklmnopqrstuvwxyz()[]{}|&%@#0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	toAscii: func (alphabet: String) -> String {
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Is there a reason you're not doing this with `Text`?

File: source/draw/RasterMonochrome.ooc [hunk 9]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		for (i in 0 .. 128)
+			alphabetMap[i] = 0
+		for (i in 0 .. alphabetSize) {
+			code := alphabet[i] as Int
+			raise(code < 32 || code > 126, "Character '" + alphabet[i] + "' (" + code toString() + ") is not printable standard ascii!")
+			raise(alphabetMap[code] > 0, "Character '" + alphabet[i] + "' (" + code toString() + ") is used more than once!")
+			value := ((i as Float) * (255.0f / ((alphabetSize - 1) as Float))) as Int clamp(0, 255)
+			alphabetMap[code] = value
+		}
+		// Allocate image
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Not needed

File: source/draw/RasterMonochrome.ooc [hunk 10]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		raise(width <= 0 || height <= 0, "An ascii image had zero dimensions!")
+		result := This new(IntVector2D new(width, height))
+		// Fill pixels
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Not needed

File: source/draw/RasterMonochrome.ooc [hunk 11]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+		x = 0
+		y = -1
+		inLine = false
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Is there a more descriptive name?

File: source/draw/RasterMonochrome.ooc [hunk 12]
```diff
@@ -224,4 +224,107 @@ RasterMonochrome: class extends RasterPacked {
...
+	fromAscii: static func (content: String) -> This {
+		// Measure dimensions and read the alphabet
+		alphabet: Char[128]
+		alphabetSize := 0
+		x := 0
+		y := -1
+		width := 0
+		inLine := false
+		for (i in 0 .. (content size)) {
+			c := content[i]
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Also, I'd prefer `current` over `c` everywhere.

File: source/draw/RasterMonochrome.ooc (hunk 11 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: `quoted`?

Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: That's better. :)

File: source/draw/RasterMonochrome.ooc (hunk 3 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: It crashed somehow.

Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: What if you declare `c: Char` outside?

File: source/draw/RasterMonochrome.ooc (hunk 8 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: It just felt like the simplest solution.

Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: `TextBuilder` can append a `Char` as well. :)

Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: So I should free alphabet if the call receiver owns it?
What do I do with the return value?

Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Ah we can look at it later if someone cares, let's go with `String` right now as you've done.

**Comment by davidpiuva** [comment]
Updated

File: source/draw/RasterMonochrome.ooc [hunk 13]
```diff
@@ -224,4 +224,100 @@ RasterMonochrome: class extends RasterPacked {
...
+				quoted = true
+			i += 1
+		}
+		raise(alphabetSize < 2, "The alphabet needs at least two characters!")
+		height = y
+		raise(x > 0, "All hexadecimal images must end with a linebreak!")
+		// Create alphabet mapping from character to luma
+		alphabetMap: Byte[128]
+		for (i in 0 .. 128)
+			alphabetMap[i] = 0
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: I believe arrays are `calloc`d and therefore this is not needed.. @fredrikbryntesson @thomasfanell ?

File: source/draw/RasterMonochrome.ooc [hunk 14]
```diff
@@ -224,4 +224,100 @@ RasterMonochrome: class extends RasterPacked {
...
+		for (i in 0 .. alphabetSize) {
+			code := alphabet[i] as Int
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: Do you need to cast?

Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: Yes, arrays cannot be indexed by Char.

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: This is a stack allocation so it depends on if the C compiler.

File: source/draw/RasterMonochrome.ooc [hunk 15]
```diff
@@ -224,4 +224,100 @@ RasterMonochrome: class extends RasterPacked {
...
+		alphabetMap: Char[256]
+		scale := ((alphabet size - 1) as Float) / 255.0f
+		output := 0.49f
+		for (rawValue in 0 .. 256) {
+			alphabetMap[rawValue] = alphabet[(output as Int) clamp(0, alphabet size - 1)]
+			output += scale
+		}
+
+		result append('<') . append(alphabet) . append('>') . append('\n')
+
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: This empty line looks awkward now.

File: source/draw/RasterMonochrome.ooc [hunk 16]
```diff
@@ -224,4 +224,100 @@ RasterMonochrome: class extends RasterPacked {
...
+			code := alphabet[i] as Int
+			raise(code < 32 || code > 126, "Character '" + alphabet[i] + "' (" + code toString() + ") is not printable standard ascii!")
+			raise(alphabetMap[code] > 0, "Character '" + alphabet[i] + "' (" + code toString() + ") is used more than once!")
+			value := ((i as Float) * (255.0f / ((alphabetSize - 1) as Float))) as Int clamp(0, 255)
+			alphabetMap[code] = value
+		}
+
+		raise(width <= 0 || height <= 0, "An ascii image had zero dimensions!")
+		result := This new(IntVector2D new(width, height))
+
```
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: We can go without this line too

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (thomasfanell) [inline source/draw/RasterMonochrome.ooc]: @marcusnaslund correct

File: source/draw/RasterMonochrome.ooc (hunk 15 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: Yes, that is why I like comments for grouping sections.

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (davidpiuva) [inline source/draw/RasterMonochrome.ooc]: The generated C code has no implicit initialization before my loop.
```c
    Numbers__Byte alphabetMap[128];
    {
        Numbers__Int i;
        for (i = 0; i < 128; i++) {
            alphabetMap[i] = 0;
        }
    }
```

Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: @davidpiuva wins this round.

File: source/draw/RasterMonochrome.ooc (hunk 15 above)
Comment (marcusnaslund) [inline source/draw/RasterMonochrome.ooc]: And I like consistent levels of comments. :)

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (fredrikbryntesson) [inline source/draw/RasterMonochrome.ooc]: K.O

**Comment by marcusnaslund** [comment]
You can move back to final review then the memory leak is fixed.

**Comment by davidpiuva** [comment]
I fixed the leaks that I could detect locally with my own settings but PRInt is down now.

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (sebastianbaginski) [inline source/draw/RasterMonochrome.ooc]: > I believe arrays are callocd and therefore this is not needed

Yes, but only if allocated with operator `new`:
```ooc
zeroInitialized := Int[10] new()
garbageInitialized: Int[10]
```
:)

**Comment by davidpiuva** [comment]
PRInt is running again.

File: source/draw/RasterMonochrome.ooc (hunk 13 above)
Comment (thomasfanell) [inline source/draw/RasterMonochrome.ooc]: Yeah for some reason I missed that we were dealing with a C array

**Comment by davidpiuva** [comment]
@marcusnaslund All checks passed.

File: test/draw/RasterMonochromeTest.ooc [hunk 17]
```diff
@@ -45,6 +45,14 @@ RasterMonochromeTest: class extends Fixture {
 			expect(image1 equals(image2))
 			image1 referenceCount decrease(); image2 referenceCount decrease()
 		})
+		this add("lossy ascii serialization", func {
+			image1 := RasterMonochrome open(this sourceSpace)
+			alphabet := " .,-_':;!+~=^?*abcdefghijklmnopqrstuvwxyz()[]{}|&%@#0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+			asciiImage := image1 toAscii(alphabet)
+			image2 := RasterMonochrome fromAscii(asciiImage)
+			expect(image1 distance(image2) < 0.01)
```
Comment (marcusnaslund) [inline test/draw/RasterMonochromeTest.ooc]: Sorry, just one more thing.
`expect(image1 distance(image2), is lessThan(0.01f))` (note that `distance` returns a `Float` as well)

Comment (davidpiuva) [inline test/draw/RasterMonochromeTest.ooc]: Updated

**Comment by davidpiuva** [comment]
All checks passed again.

**Pull request merged**